### PR TITLE
main branch for refactoring and bugfixes after PSE

### PR DESCRIPTION
**Pull request by jbessi**

### CODE REVIEW

File: src/Components/ButtonGroup/ButtonGroup.tsx [hunk 1]
```diff
@@ -7,12 +7,32 @@ import convertDate from "../../Services/convert-date.service";
 import RouterService from "../../Services/router.service";
 import MobileButton from "./MobileButton";
 import ConvertDateService from "../../Services/convert-date.service";
+import {IVersions, IButtonLabels, IButtonGroup} from "../../interfaces";
```
Comment (tschimbr) [inline src/Components/ButtonGroup/ButtonGroup.tsx:10]: Move these interfaces to this file as well.

File: src/Services/convert-date.service.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
+/**
+ * Convert the date from a text to a integer and gives it back as a date string
+ * @param value
+ * @returns {*}
+ * @constructor
+ */
+function ConvertDateService(value) {
```
Comment (tschimbr) [inline src/Services/convert-date.service.tsx:7]: There are probably date parsing and printing methods already implemented in standard libraries. I know this is not part of this PR, but we should do this later on.

Comment (tschimbr) [reply src/Services/convert-date.service.tsx:7]: This file should not be necessary. Maybe also just use another date format in the datepickers.

File: src/Utils.tsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
+export const initialCodeState = {
+    attributes: {
+        code: "",
+        text: "",
+        parent: null,
```
Comment (tschimbr) [inline src/Utils.tsx:5]: Are all the null entries needed? The attributes should be optional.

File: src/__test__/breadcrumbstest.spec.tsx [hunk 4]
```diff
@@ -1,13 +1,14 @@
-const { Builder, By} = require('selenium-webdriver')
-const {browser, sleep, n, options} = require("../setupTests");
-
 describe('Default Suite', function() {
   let driverCrumbs
+  var webdriver = require("selenium-webdriver");
```
Comment (tschimbr) [inline src/__test__/breadcrumbstest.spec.tsx:3]: Have a look at our feature test configuration for Casematch: https://gitlab.eonum.ch/eonum/casematch/-/blob/master/spec/rails_helper.rb

Maybe there are other more suitable webdrivers and/or options.

Comment (tschimbr) [reply src/__test__/breadcrumbstest.spec.tsx:3]: We use chromium in Casematch instead of Firefox.

File: src/__test__/breadcrumbstest.spec.tsx [hunk 5]
```diff
@@ -1,13 +1,14 @@
...
+  var By = webdriver.By;
+  var {browser, sleep, n, options} = require("../setupTests");
+
   beforeAll( async function() {
     await sleep(n);
   })
   beforeEach(async function() {
-    driverCrumbs = await new Builder().forBrowser(browser).setFirefoxOptions(options).build();
+    driverCrumbs = new webdriver.Builder().forBrowser(browser).setFirefoxOptions(options).build()
```
Comment (tschimbr) [inline src/__test__/breadcrumbstest.spec.tsx:11]: I am not sure if this is the right place. We should not have to instantiate a new webdriver for every test. This slows down test execution a lot. However I am not sure what's best practice.

Comment (tschimbr) [reply src/__test__/breadcrumbstest.spec.tsx:11]: But first we'll have to fix all tests before we refactor.

File: src/__test__/search.spec.tsx [hunk 6]
```diff
@@ -1,15 +1,17 @@
-const { Builder, By, Key, until } = require('selenium-webdriver')
-const assert = require('assert')
-const {urlMatches} = require("selenium-webdriver/lib/until");
-const {browser, sleep, n, options} = require("../setupTests");
-
 describe('search', function() {
   let driverSearch;
+  var webdriver = require("selenium-webdriver");
```
Comment (tschimbr) [inline src/__test__/search.spec.tsx:3]: And there is awful lot of code duplication..

**Review by tschimbr — COMMENTED** [review]
LGTM. There is a lot to be done, but not regarding typescript migration. I'd just move all interfaces used only in one place to the place they are used.

File: src/Services/convert-date.service.tsx (hunk 2 above)
Comment (tschimbr) [reply src/Services/convert-date.service.tsx:7]: And currently the American date format is used which is confusing. We should use a Swiss / European format.

File: src/Components/Bodies/CodeBodyUnversionized.tsx [hunk 7]
```diff
@@ -194,42 +198,42 @@ class BodyII extends Component {
...
-            categories.push(
+        if(this.state.attributes["children"] && this.state.attributes["children"].length > 0) {
+            attributes_html.push(
                 <div key={i}>
                     <h5>{translateJson["LBL_CHILDREN"]}</h5>
                     <ul>
-                        {this.state.categories["children"].map((child, i) => (
+                        {this.state.attributes["children"].map((child, i) => (
                             <li key={i}><a key={"link to child: " + i} className="link" onClick={() => {this.goToChild(child)}}>{child.code}: </a>
                                 <span key={"child text"} dangerouslySetInnerHTML={{__html: child.text}}/></li>
```
[the remark this answers is not in the document]
Comment (jbessi) [reply src/Components/Bodies/CodeBodyUnversionized.tsx:229]: In Migel there are some, but just 8 out of 1325...

**Comment by jbessi** [comment]
Working state after TS migration and test refactoring (changed testframework to jest-puppeteer and adding / deleting / rewriting tests).

**Pull request merged**